src/sleep/sleep.c

```diff
@@ -44,49 +48,107 @@
 
 static SleepOperation arg_operation = _SLEEP_OPERATION_INVALID;
 
-static int write_hibernate_location_info(const HibernateLocation *hibernate_location) {
+static int write_efi_hibernate_info(const char *device, const char *offset, int log_level) {
+        int r = 0;
+
+#if ENABLE_EFI_HIBERNATE_INFO
+        _cleanup_(blkid_free_probep) blkid_probe b = NULL;
+        _cleanup_free_ char *info = NULL;
+        char info_name[STRLEN("HibernateInfo-") + SD_ID128_STRING_MAX + STRLEN("-" EFI_VENDOR_SYSTEMD_STR)];
+        const char *uuid;
+        sd_id128_t machine_id = SD_ID128_NULL;
+
+        if (!is_efi_boot())
+                return 0;
+
+        errno = 0;
+        b = blkid_new_probe_from_filename(device);
+        if (!b)
+                return log_full_errno(log_level,
+                                      errno_or_else(ENOMEM),
+                                      "Failed to create blkid probe for device '%s': %m", device);
+
+        blkid_probe_enable_superblocks(b, true);
+        blkid_probe_set_superblocks_flags(b, BLKID_SUBLKS_UUID);
+
+        errno = 0;
+        r = blkid_do_safeprobe(b);
+        if (IN_SET(r, _BLKID_SAFEPROBE_ERROR, _BLKID_SAFEPROBE_AMBIGUOUS, _BLKID_SAFEPROBE_NOT_FOUND))
+                return log_full_errno(log_level,
+                                      errno_or_else(EIO),
+                                      "Failed to probe superblock of device '%s': %m", device);
+        assert(r == _BLKID_SAFEPROBE_FOUND);
+
+        r = blkid_probe_lookup_value(b, "UUID", &uuid, NULL);
+        if (r < 0)
+                return log_full_errno(log_level,
+                                      errno_or_else(EIO),
+                                      "Failed to probe filesystem UUID of device '%s': %m", device);
+
+        r = asprintf(&info, "UUID=%s:%s", uuid, offset);
+        if (r < 0)
+                return log_oom();
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0 && !ERRNO_IS_MACHINE_ID_UNSET(r))
+                return log_full_errno(log_level, r, "Failed to get machine-id: %m");
+
+        if (!sd_id128_is_null(machine_id))
+                xsprintf(info_name,
+                         "HibernateInfo-" SD_ID128_FORMAT_STR "-" EFI_VENDOR_SYSTEMD_STR,
+                         SD_ID128_FORMAT_VAL(machine_id));
+        else
+                strcpy(info_name, "HibernateInfo" "-" EFI_VENDOR_SYSTEMD_STR);
```

Comment:
btw, I think we should be more precise here, and call this HibernateLocation or so, because that's really what it is about